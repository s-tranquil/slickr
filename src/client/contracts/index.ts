export type { IRecentPictureCollection } from "./IRecentPictureCollection";
export type { IRecentPicture } from "./IRecentPicture";
