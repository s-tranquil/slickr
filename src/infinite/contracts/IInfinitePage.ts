interface IInfinitePage<TItem> {
    items: TItem[];
    totalPages: number;
}

export type { IInfinitePage };
