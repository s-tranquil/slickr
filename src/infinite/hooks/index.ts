export { useIsScrollBottom } from "./useIsScrollBottom";
export { useWindowChangeListener } from "./useWindowChangeListener";
