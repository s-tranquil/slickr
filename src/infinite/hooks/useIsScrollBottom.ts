import { useCallback } from "react";

// determines if window scroll is at the bottom of the page
const useIsScrollBottom = () => useCallback(
    () => {
        const windowHeight = "innerHeight" in window ? window.innerHeight : document.documentElement.offsetHeight;
        const body = document.body;
        const html = document.documentElement;
        const docHeight = Math.max(
            body.scrollHeight,
            body.offsetHeight,
            html.clientHeight,
            html.scrollHeight,
            html.offsetHeight
        );
        const windowBottom = windowHeight + window.pageYOffset;

        return  windowBottom >= docHeight - 1;
    },
    []
);

export { useIsScrollBottom };
