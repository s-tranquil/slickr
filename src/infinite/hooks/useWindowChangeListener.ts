import {
    useEffect,
    useRef
} from "react";

type HandlerType = () => void;

// calls handler on window scroll and resize events
const useWindowChangeListener = (handler: HandlerType) => {
    const savedHandler = useRef<HandlerType>();

    useEffect(
        () => {
            savedHandler.current = handler;
        }, 
        [handler]
    );

    useEffect(
        () => {
            const eventListener = () => (savedHandler.current as HandlerType)();
            window.addEventListener("scroll", eventListener);
            window.addEventListener("resize", eventListener);
    
            return () => {
                window.removeEventListener("scroll", eventListener);
                window.removeEventListener("resize", eventListener);
            };
        },
        []
    );
}

export { useWindowChangeListener };
