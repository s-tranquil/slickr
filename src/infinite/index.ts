export { Infinite } from "./components/Infinite";
export type { IInfinitePage } from "./contracts/IInfinitePage";