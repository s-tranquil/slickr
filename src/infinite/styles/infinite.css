.infinite__loader {
    width: 100%;
    height: 200px;
    position: absolute;
    top: 30%;
}

.infinite__items {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(240px, 1fr));
    grid-gap: 0.5rem;
    height: 100%;
}
