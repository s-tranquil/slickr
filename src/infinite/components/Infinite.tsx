import React from "react";

import "../styles/infinite.css";

import {
    useIsScrollBottom,
    useWindowChangeListener
} from "infinite/hooks";
import { Loader } from "loader";

import type { IInfinitePage } from "../contracts/IInfinitePage";
interface IProps<TItem> {
    fetchData: (pageNo: number) => Promise<IInfinitePage<TItem>>;
    renderItem: (item: TItem) => JSX.Element;
}

function Infinite<TItem>({
    renderItem,
    fetchData
}: IProps<TItem>) {
    // TODO: switch to reducer
    const [items, setItems] = React.useState<TItem[]>([]);
    const [isLoading, setIsLoading] = React.useState<boolean>(false);
    const [pageNo, setPageNo] = React.useState<number>(0);
    const [totalPages, setTotalPages] = React.useState<number>(1);

    const fetchPage = React.useCallback(
        async (pageNo: number) => {
            setIsLoading(true);

            const data = await fetchData(pageNo);

            setItems(oldItems => oldItems.concat(data.items));
            setTotalPages(data.totalPages);

            setPageNo(pageNo);
            setIsLoading(false);
        },
        [fetchData]
    );

    const fetchNextPage = React.useCallback(
        async () => {
            await fetchPage(pageNo + 1);
        },
        [fetchPage, pageNo]
    );

    const isScrollBottom = useIsScrollBottom();
    const needsNextPage = React.useCallback(
        () => {
            return (
                !isLoading &&
                isScrollBottom() &&
                pageNo < totalPages
            );
        },
        [pageNo, totalPages, isScrollBottom, isLoading]
    );

    const getNextPageIfNeeded = React.useCallback(
        async () => {
            if (needsNextPage()) {
                await fetchNextPage();
            }
        },
        [fetchNextPage, needsNextPage]
    );

    const onScroll = React.useCallback(getNextPageIfNeeded, [getNextPageIfNeeded]);
    useWindowChangeListener(onScroll);

    // check if we need to load more after previous page loaded, and for the first page
    React.useEffect(
        () => {
            if (!isLoading) {
                const fetchNextPageAsync = async () => {
                    await getNextPageIfNeeded()
                };
                fetchNextPageAsync();
            }
        },
        [getNextPageIfNeeded, isLoading]
    );

    return (
        <div className="infinite">
            {!items.length && (
                <div className="infinite__loader infinite__loader_initial">
                    <Loader />
                </div>
            )}
            {!!items.length && (
                <>
                    <div className="infinite__items">
                        {items.map((item: TItem, index: number) => (
                            // it's okay to use index as a key as we only extend items array
                            <InfiniteItem
                                key={index}
                                item={item}
                                renderItem={renderItem}
                            />
                        ))}
                    </div>
                    {pageNo < totalPages && (
                        <div className="infinite__loader infinite__loader_more">
                            <Loader />
                        </div>
                    )}
                </>
            )}
        </div>
    );
}

// It is an empty wrapper
// Needed because if replaced with <React.Fragment> stops working in production build
function InfiniteItem<T>(
    {
       item,
       renderItem
    }: {
        item: T,
        renderItem: (item: T) => React.ReactElement
    }
) {
    return (<>{renderItem(item)}</>);
}

export { Infinite };
