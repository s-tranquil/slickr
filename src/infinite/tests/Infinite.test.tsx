import React, { Dispatch } from "react";

import {
    mount,
    shallow
} from "enzyme";
import { IInfinitePage } from "infinite/contracts/IInfinitePage";
import * as infiniteHooks from "infinite/hooks";

import { Infinite } from "../components/Infinite";

interface ITestData {
    text: string;
}

const renderInfinite = (fetchData: (pageNo: number) => Promise<IInfinitePage<ITestData>>) => (
    <Infinite
        fetchData={fetchData}
        renderItem={(item: ITestData) => (<div key={item.text}>{item.text}</div>)}
    />
);

const renderEmpty = () => {
    const emptyDataMock = jest.fn().mockResolvedValue([] as ITestData[]);
    return renderInfinite(emptyDataMock);
}

describe("Infinite", () => {
    beforeEach(() => {
        jest.spyOn(React, "useCallback").mockImplementation(callback => (...args) => callback(args));
        jest.spyOn(React, "useState").mockImplementation(((initialState: unknown) => {
            let state: unknown = initialState;
            const setState: Dispatch<unknown> = (newState) => state = newState;
            return [state, setState] as [unknown, Dispatch<unknown>];
        }) as any);
        jest.spyOn(React, "useEffect").mockImplementation(_ => {});
        jest.mock("infinite/hooks", () => ({
            useIsScrollBottom: jest.fn(() => () => true),
            useWindowChangeListener: jest.fn((_) => {})
        }));
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it("shows initial loader when no data present", () => {
        const wrapper = shallow(renderEmpty());
        const loader = wrapper.find("div.infinite__loader_initial");
        expect(loader.exists()).toBeTruthy();
    });

    it("matches snapshot when no data present", () => {
        const wrap = mount(renderEmpty());
        expect(wrap).toMatchSnapshot();
    });

    it("matches snapshot when one page passed", async () => {
        jest.spyOn(React, "useEffect").mockImplementation(f => f());
        // jest.spyOn(React, "useCallback").mockImplementation(f => f());
        const dataMock = jest.fn().mockResolvedValue({
            items: [
                { text: "1" },
                { text: "2" }
            ],
            totalPages: 1
        } as IInfinitePage<ITestData>);
        const wrap = await mount(renderInfinite(dataMock));
        wrap.update();
        expect(wrap).toMatchSnapshot();
        expect(dataMock).toBeCalledTimes(1);
    });
});