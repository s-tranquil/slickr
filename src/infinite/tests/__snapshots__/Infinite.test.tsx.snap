// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infinite matches snapshot when no data present 1`] = `
<Infinite
  fetchData={[MockFunction]}
  renderItem={[Function]}
>
  <div
    className="infinite"
  >
    <div
      className="infinite__loader infinite__loader_initial"
    >
      <Loader>
        <div
          className="loader"
        >
          <svg
            className="loader__icon"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 24 30"
            x="0px"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
            y="0px"
          >
            <rect
              fill="#333"
              height="5"
              width="4"
              x="0"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
            <rect
              fill="#333"
              height="5"
              width="4"
              x="10"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0.15s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0.15s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
            <rect
              fill="#333"
              height="5"
              width="4"
              x="20"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0.3s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0.3s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
          </svg>
        </div>
      </Loader>
    </div>
  </div>
</Infinite>
`;

exports[`Infinite matches snapshot when one page passed 1`] = `
<Infinite
  fetchData={
    [MockFunction] {
      "calls": Array [
        Array [
          Array [
            1,
          ],
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  renderItem={[Function]}
>
  <div
    className="infinite"
  >
    <div
      className="infinite__loader infinite__loader_initial"
    >
      <Loader>
        <div
          className="loader"
        >
          <svg
            className="loader__icon"
            id="Layer_1"
            version="1.1"
            viewBox="0 0 24 30"
            x="0px"
            xmlSpace="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
            y="0px"
          >
            <rect
              fill="#333"
              height="5"
              width="4"
              x="0"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
            <rect
              fill="#333"
              height="5"
              width="4"
              x="10"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0.15s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0.15s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
            <rect
              fill="#333"
              height="5"
              width="4"
              x="20"
              y="13"
            >
              <animate
                attributeName="height"
                attributeType="XML"
                begin="0.3s"
                dur="0.6s"
                repeatCount="indefinite"
                values="5;21;5"
              />
              <animate
                attributeName="y"
                attributeType="XML"
                begin="0.3s"
                dur="0.6s"
                repeatCount="indefinite"
                values="13; 5; 13"
              />
            </rect>
          </svg>
        </div>
      </Loader>
    </div>
  </div>
</Infinite>
`;
